### PR TITLE
[mac] use link layer sequnece number to detect duplicate frames

### PR DESCRIPTION
**Pull request by zhanglongxia**
Openthread is using the frame counter in security header to detect
duplicate frames and it also provides API to enable/disable link layer
security. If link layer security is disabled, the MAC layer can't detect
duplicate frames any more. This patch resolves this issue by using the
link layer sequence number to detect duplicate frames.

### CODE REVIEW

**Comment by abtink** [comment]
@zhanglongxia , thanks for doing this. 

How can we handle when a device/child gets reset/rebooted? It starts from zero sequence number but its previous sequence number is remembered by all the neighbors/parent in the network and I guess they may reject new frames from the device after reset when checking the sequence number.

With regards to security MLE and MAC frame counters, we save them in flash/settings with a jump ahead increase `OPENTHREAD_CONFIG_STORE_FRAME_COUNTER_AHEAD` to ensure that on a reset we start after the previous counter value. They are `uint32` and never wrap around.

File: src/core/mac/mac.cpp [hunk 1]
```diff
@@ -1595,6 +1595,26 @@ void Mac::HandleReceiveTimer(void)
...
 }
 
+otError Mac::ProcessFrameSequence(Frame &aFrame, const Address &aSrcAddr)
+{
+    otError   error        = OT_ERROR_NONE;
+    Neighbor *linkNeighbor = GetNetif().GetMle().GetLinkNeighbor(aSrcAddr);
+
+    VerifyOrExit(linkNeighbor != NULL);
+
+    if (aFrame.GetSequence() + 1 <= linkNeighbor->GetLinkFrameSequence())
```
Comment (abtink) [inline src/core/mac/mac.cpp]: How to handle `uint8_t` wrap around and possibly missing frames ?

For example, if `GetLinkFrameSequence()` is at `0xff`, and we miss a frame (miss receiving sequence number `0xff` itself), then we reject the next frame with seq number `0` and all the next ones (which are all technically ahead/after sequence number `0xff`)

**Comment by zhanglongxia** [comment]
@abtink  Current code can’t handle the case you mentioned. When the router/child gets reset/rebooted, router/child will send LinkRequest/ChildUpdateRequest messages to neighbors/parent to recover its previous status. After the neighbors/parent receives these messages, it knows that this router/child is coming out of reset/reboot. Then the neighbors/parent could initial the link sequence again. If I update the code like this, could it handle reset/reboot cases? 

As for the uint8_t wrap round issue, could we only detect the duplicate frames and update the code like this?
    if (aFrame.GetSequence() + 1 == linkNeighbor->GetLinkFrameSequence())
    {
        ExitNow(error = OT_ERROR_DUPLICATED);
    } 
    else 
    {
        linkNeighbor->SetLinkFrameSequence(aFrame.GetSequence() + 1);
    }

**Comment by abtink** [comment]
@zhanglongxia , thanks for explanation.

About multicast msg do we not use encryption? (so that we can re-use the security counter for dup detection)? 

About multiple MLE parent request/response, how can the seq number duplicate detection work/help for new device during attach? To handle device/child reset/reboot correctly, we need to accept any seq number for a "Parent Request" message.

For multiple MLE Parent Request, I would suggest  a solution at MLE layer on how we process the parent requests to either avoid generating the response in first place or reuse the same challenge in response for a second request. For example, on parent side we can see if there is already an MLE response in queue for same child and remove it from the queue if we receive a new parent request from same child. Or ensure to re-use the same challenge for the child if the child is already in `kStateParentRequest` state when we receive a second "Parent Request".

**Comment by zhanglongxia** [comment]
@abtink  Thanks for you good suggestion.

The multicast msg uses encryption. We can use the security counter for dup detection.

When parent receives the first MLE parent request, it will initialize the link sequence number at the same time. When the second MLE parent request msg with the same sequence number comes，it will be dropped by MAC layer.

I just tried the methods you suggested, both of them work well. So this commit is not necessary any more, I will close it.

**Pull request closed without merging**